src/cascadia/TerminalApp/IPaneContent.idl

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+namespace TerminalApp
+{
+
+    runtimeclass BellEventArgs
+    {
+        Boolean FlashTaskbar { get; };
+    };
+
+    interface IPaneContent
+    {
+        Windows.UI.Xaml.FrameworkElement GetRoot();
+
+        Windows.Foundation.Size MinSize { get; };
+
+        String Title { get; };
+        UInt64 TaskbarState { get; };
+        UInt64 TaskbarProgress { get; };
+        Boolean ReadOnly { get; };
+
+        Microsoft.Terminal.Settings.Model.NewTerminalArgs GetNewTerminalArgs(Boolean asContent);
+
+        void Focus(Windows.UI.Xaml.FocusState reason);
+
+        void Close();
+
+        event Windows.Foundation.TypedEventHandler<Object, Object> CloseRequested;
+
+        event Windows.Foundation.TypedEventHandler<Object, BellEventArgs> BellRequested;
+        event Windows.Foundation.TypedEventHandler<Object, Object> TitleChanged;
+        event Windows.Foundation.TypedEventHandler<Object, Object> TabColorChanged;
+        event Windows.Foundation.TypedEventHandler<Object, Object> TaskbarProgressChanged;
+        event Windows.Foundation.TypedEventHandler<Object, Object> ConnectionStateChanged;
+        event Windows.Foundation.TypedEventHandler<Object, Object> ReadOnlyChanged;
+        event Windows.Foundation.TypedEventHandler<Object, Object> FocusRequested;
+    };
+
+
+    enum PaneSnapDirection
+    {
+        Width,
+        Height
+    };
+
+    interface ISnappable
+    {
+        Single SnapDownToGrid(PaneSnapDirection direction, Single sizeToSnap);
+        Windows.Foundation.Size GridSize { get; };
+    };

```

Comment:
What panes aren't snappable?

FWIW, if we do follow common OOP design guidelines, asking for whether something implements X and then taking action upon that without actually using the interface is a violation of the encapsulation principle. It leaks the behaviors that should be internal to the interface implementation to outside of it. I'm not entirely sure what name this anti-pattern has, but calling it an [Anemic domain model](https://en.wikipedia.org/wiki/Anemic_domain_model) is probably fair.

Reply:
I mean, like, a settings pane wouldn't really be snappable, right? Like, it's `GridSize` is just 1x1. 

I don't think I'm asking anywhere if a `PaneContent` is an `ISnappable`, then _not_ using the `ISnappable`.
* In `Pane::_CalcSnappedDimension` I'm checking if it's snappable. If it is, we try to snap up/down to the grid. Otherwise we just return the requested size
* In `Pane::_AdvanceSnappedDimension`, if it's not snappable we just add 1, otherwise we use the grid size from the snappable interface

I dunno, this _seemed_ like a viable abstraction. I'm focusing here on the fact we're gonna have 3p folks try and author `IPaneContent`. Most of the time, I don't think they'll need to implement snapping.